backtesting/backtesting.py

```diff
@@ -1221,101 +1237,154 @@ def maximize(stats: pd.Series, _key=maximize):
+                else:
+                    if os.name == 'posix':
+                        warnings.warn("For multiprocessing support in `Backtest.optimize()` "
+                                      "set multiprocessing start method to 'fork'.")
+                    for batch_index in _tqdm(range(len(param_batches))):
+                        _, values = Backtest._mp_task(backtest_uuid, batch_index)
                         for value, params in zip(values, param_batches[batch_index]):
                             heatmap[tuple(params.values())] = value
+            finally:
+                del Backtest._mp_backtests[backtest_uuid]
+
+            best_params = heatmap.idxmax()
+
+            if pd.isnull(best_params):
+                # No trade was made in any of the runs. Just make a random
+                # run so we get some, if empty, results
+                self.run(**param_combos[0])  # type: ignore
             else:
-                if os.name == 'posix':
-                    warnings.warn("For multiprocessing support in `Backtest.optimize()` "
-                                  "set multiprocessing start method to 'fork'.")
-                for batch_index in _tqdm(range(len(param_batches))):
-                    _, values = Backtest._mp_task(backtest_uuid, batch_index)
-                    for value, params in zip(values, param_batches[batch_index]):
-                        heatmap[tuple(params.values())] = value
-        finally:
-            del Backtest._mp_backtests[backtest_uuid]
-
-        best_params = heatmap.idxmax()
-
-        if pd.isnull(best_params):
-            # No trade was made in any of the runs. Just make a random
-            # run so we get some, if empty, results
-            self.run(**param_combos[0])  # type: ignore
+                # Re-run best strategy so that the next .plot() call will render it
+                self.run(**dict(zip(heatmap.index.names, best_params)))
+
+            if return_heatmap:
+                return self._results, heatmap
+            return self._results
+
+        def _optimize_skopt() -> Union[pd.Series, Tuple[pd.Series, pd.Series]]:
+            """
+            Optimize strategy parameters to an optimal combination using
+            scikit-optimize. Returns result `pd.Series` of
+            the best run.
+
+            skopt.forest_minimize(), a tree based regression model  is used
+            to model the expensive to evaluate function (Backtest.run()). The
+            model is improved by sequentially evaluating the expensive function
+            at the next best point. Thereby finding the minimum of func with as
+            few evaluations as possible: 
+            (https://scikit-optimize.github.io/stable/modules/generated/skopt.forest_minimize.html).
+
+            The default value for base_estimator is used (ET), and n_calls = max_tries 
+
+            .. TODO::
+                Improve multiprocessing with joblib.parallel .
```

Comment:
Good point. skopt.forest_minimize() has a number of input parameters that we do not include Backtest.optimize(), I have only refenced the best estimator function that we use ET by default. In a future release maybe more refined input for optimization can be allowed. For the n_jobs, -1 uses the maximum CPU which I think is the best default.

Reply:
I was contemplating why you went with `forest_minimize()` (and its ET) over `gp_minimize()` or `gbrt_minimize()`. Welcome your thoughts/insights.

> For the n_jobs, -1 uses the maximum CPU which I think is the best default.

I tentatively agree, but spawning multiple processes is [not without overhead](https://stackoverflow.com/a/32649856/1090455). I'm further worried how joblib does memory management of closure structures. If it's not copy-on-write, this will spell disaster for larger data payloads. And with pickling, `n_jobs != 1` might not even work! I know I've went through some hoops to get multiprocessing working well (https://github.com/kernc/backtesting.py/commit/1faa7e863cd11ff7a26fb7f6bd445095978019ee).

Reply:
Sorry, I probably screwed it up. I guess I wasn't supposed to push the egg-info folder? Didn't expect would add so many files.

Reply:
Yeah, I use a global gitignore file with patterns for common build artifacts for all languages I touch, so I don't even notice when a project's gitignore is amiss. :sweat_smile:

This one seems to exclude some common dir's contents, so it should exclude the dirs as well. :thinking: Guess not.
https://github.com/kernc/backtesting.py/blob/252cabdf15fd8827560393cc4eb64d48a5a66bd9/.gitignore#L6-L9

Reply:
My trials with `n_jobs=-1` were all running slower. `n_jobs=` in scikit-optimize is only used for building the hyperspace landscape estimators, not for running the expensive objective function. :disappointed: